src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,160 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            addFirst(value);
+        } else {
+            addLast(value);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddIndex(index);
+        if (head == null) {
+            addFirst(value);
+        } else if (size == index) {
+            addLast(value);
+        } else {
+            Node<T> currentIndexNode = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(currentIndexNode.prev, value, currentIndexNode);
+            if (currentIndexNode.prev != null) {
+                currentIndexNode.prev.next = newNode;
+            } else {
+                head = newNode;
+            }
+
+            currentIndexNode.prev = newNode;
+        }
+        size++;
+    }
+
+    private void addFirst(T value) {
+        Node<T> nodeToAdd = new Node<>(null, value, null);
+        head = nodeToAdd;
+        tail = nodeToAdd;
+    }
+
+    private void addLast(T value) {
+        Node<T> nodeToAdd = new Node<>(tail, value, null);
+        tail.next = nodeToAdd;
+        tail = nodeToAdd;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldItem = node.item;
+        node.item = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        unlink(node);
+        size--;
```

Comment:
The size is decremented after unlinking the node, which is correct. Ensure that this logic is consistently applied across all removal methods.